[FSDP] Add rate limiter

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #84911 [FSDP] Add `use_orig_params`
* #84910 [FSDP] Post-bwd prefetch with reverse pre-forward order
* **#83917 [FSDP] Add rate limiter**
* #84366 [FSDP][Easy] Save unpadded/padded unsharded sizes as attributes
* #83665 [FSDP] Generalize prefetching; lower unshard/reshard to handle
* #84761 [FSDP][Easy] Minor cleanup
* #84601 [FSDP] Subtest prefetching for `test_fsdp_grad_acc.py`
* #84600 [FSDP] Remove `forward_prefetch`

**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

Commit:
[FSDP] Add rate limiter

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

awgu commented:
I messed up my local git, so the original rate limiter PR got detached from the stack.
re: https://github.com/pytorch/pytorch/pull/83666#pullrequestreview-1081612663

> Looks good overall, but I guess the API is still pending discussions? A couple of suggestions:
> 
> 1. I see the initial experimentation on T5, can we find the set of representative workloads where we saw high GPU memory issue and evaluate impact of this diff on those as well?
> 2. Maybe full_param_size is a better metric to ratelimit on, because # of all_gathers will have to be tuned if we shard differently / more / less granular?
> 3. Does this also ratelimit the backward?

1. Yup, I will have thorough experiment results once we finalize the design more.
2. Good point! I need to dig more into this and am open to discussion.
    - To me, the broader question is whether we should have oscillatory behavior or uniform behavior. For oscillatory, I mean that we may enqueue multiple before dequeueing and we may dequeue multiple before the next enqueue. For uniform, I mean that we only dequeue one at a time, so the limiter reaches a simple steady state.
    - @mrshenli Do you know if we record CUDA events `e1`, ..., `ek` sequentially, can we directly call `ek.synchronize()` to synchronize all of `e1`, ..., `ek`? Will this be more performant than calling `e1.synchronize()`, ..., `ek.synchronize()` individually?
    - At the time of this comment, the PR's implementation enqueues until it hits the a count-based limit and dequeues one per iteration thereafter. This exemplifies the "uniform behavior" I mentioned.
    - For the parameter size heuristic, I need to determine how to choose a good value. Perhaps, it should be proportional to the max GPU memory.
3. Yes, this also rate limits the backward. We can see this because the limiter records frees and synchronizes before all-gathers, both of which still happen in the backward.

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

lessw2020 commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -938,6 +946,8 @@ def __init__(
         self.cpu_offload = cpu_offload or CPUOffload()
         self.backward_prefetch = backward_prefetch
+        self.limit_all_gathers = limit_all_gathers
+        self._max_num_inflight_all_gathers = 2  # empirically chosen
```
a - I think it would be useful to expose "max_num_inflight_all_gathers" during init rather than embedding it, esp to allow larger customers to tune to their own hardware/models?  (I'm going to do this while testing this fix for example). 
b - If that is useful, then these two vars could also be compressed into a single 
self.max_num_all_gathers=2,
where 0 is Off, and thus no limit, and then 1, 2, etc. with default of 2, to simplify api and internal checks, rather than exposing two params?

awgu replied:
I had a previous version that did what you suggested (compress into one variable). I changed it since I thought it might be confusing for non-advanced users, for whom it may be better to expose a default value. For advanced users, we can expose a setter method that can be called post-construction. Of course, all of this is open for discussion. We are still thinking about the design.

I am not updating this PR at the moment as I experiment with different variations. I am planning to experiment with a size-based rule instead of a count-based rule and to flush the queue instead of only dequeuing a single event at a time. If your experiments are not urgent, I may recommend considering to wait a few days until I solidify the approach more.

lessw2020 replied:
Sounds good - I'll hold off on doing much work with this fix then re: possible move to size based.  
Thanks for the updates!

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -975,6 +1023,10 @@ def __init__(
         self._materialize_module(module, param_init_fn, device_from_device_id)
         self._move_module_to_device(module, ignored_params, device_from_device_id)
         self.compute_device = self._get_compute_device(module, ignored_params, device_from_device_id)
+        self._max_inflight_all_gather_size = (
```
@lessw2020 This is the internal knob to tune for this size-based limit.

lessw2020 replied:
got it, thanks!

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1720,6 +1795,7 @@ def _lazy_init(self) -> None:
         self._exec_order_data.init(self, self.process_group)
         # Initialize non-root FSDP instances and share attributes from the root
         # to non-root instances
+        inconsistent_limit_all_gather = False
```
TODO (nit): rename this to `inconsistent_limit_all_gathers`

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1455,6 +1508,20 @@ def _unshard(
+                    events.append(event)
+            # As an optimization, only synchronize the latest event
+            if events:
+                events[-1].cuda_event.synchronize()
```
just double confirm, when we synchronize the latest event, the previous events are synced, right? because this is a full CPU/GPU sync

awgu replied:
We may want some ask some experts to confirm this.

However, conceptually, it made sense to me that we only need to sync the latest event. By calling `.synchronize()`, we wait for that latest free to finish, at which point all earlier frees (which happen in the same stream, the default stream) must have finished as well since they are obey the sequential order ensured by the stream semantics.

I have not tested if there is any performance gain by only calling `.synchronize()` on the latest event. However, it seems that it cannot be slower compared to calling it on all of them.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -975,6 +1023,10 @@ def __init__(
         self.compute_device = self._get_compute_device(module, ignored_params, device_from_device_id)
+        self._max_inflight_all_gather_size = (
+            torch.cuda.get_device_properties(self.compute_device).total_memory
+            * 0.005  # empirically chosen, e.g. 200 MB limit for 40 GB GPU
```
assume we should choose larger limit to avoid frequently slow down the CPU thread in default, otherwise slow CPU may result in larger cuda kernel scheduling overhead and slow down training?

awgu replied:
I have been experimenting with this (but much more to come). I originally chose `0.05`, i.e. 2 GB limit, but it is basically not effective then.

The observation is that when people scale models, they add more layers, not increase the size of each layer. (For example, think of transformer layers and the transformer auto wrap policy.) This means that the size of each `FlatParameter` is actually pretty fixed when scaling the model, at least for recursive wrapping. For DeepViT, each padded unsharded flattened parameter is 16 MB. If we set a limit of 2 GB, it will take too many all-gathers to get hit, and the limit is not effective. The CPU thread will have already ran ahead.

We can keep brainstorming how to decide this limit. For this particular approach of using a scaling factor against the total GPU memory, we may be able to use the [elbow method](https://en.wikipedia.org/wiki/Elbow_method_(clustering)).

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1455,6 +1508,20 @@ def _unshard(
+        if self.limit_all_gathers:
+            queue = self._free_event_queue
+            events: List[FreeEvent] = []
+            if queue.queue_size > self._max_inflight_all_gather_size:
```
wondering whether we should use queue_size or current runtime cached/allocated memory utilization %..? if we use queue_size, the _max_inflight_all_gather_size probably needs to be larger than the max of the fsdp instances' full parameter size?

awgu replied:
> if we use queue_size, the _max_inflight_all_gather_size probably needs to be larger than the max of the fsdp instances' full parameter size?

Yes, unless you want to have only 1 inflight all-gather at a time, which I suppose should never be desired. I think until we have a better understanding of the correlation between the cached/allocated/active memory stats and the fast CPU thread, it may be hard to develop a heuristic using those.

In practice, most `FlatParameter`s should be around the same size, so approximately, a count-based limiter and a size-based limiter should be about the same. I am still working on experiments to explore.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1736,10 +1812,21 @@ def _lazy_init(self) -> None:
                 fsdp_module._need_pre_backward_unshard = self._need_pre_backward_unshard
                 for handle in fsdp_module._handles:
                     fsdp_module._init_param_attributes(handle)
+        if inconsistent_limit_all_gathers:
```
I guess it should be rarely happened? almost can not happen with autowrapping?

awgu replied:
Yes, this should not happen with auto wrapping, only with manual wrapping. @rohan-varma pointed the possibility of inconsistency out in the old version of the PR, and I think it is good to have as a safety.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1481,6 +1548,14 @@ def _reshard(
             free_unsharded_flat_params,
         ):
             handle.reshard(free_unsharded_flat_param)
```
it still calls record_stream in the reshard(), should it be removed if self.limit_all_gathers?

zhaojuanmao replied:
hmm, I guess it is fine..? because we slowed down CPU thread anyway, it is OK to let the record stream still hold the memory

awgu replied:
I think we need to still call `record_stream()` if `self.limit_all_gathers == True`. Otherwise, the memory could still be freed and reused too early (e.g. before we even get a chance to synchronize the CPU thread in `_unshard()`).

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

HamidShojanazeri commented:
Thanks @awgu  for the PR.  I just wanted to share some thoughts,  the overall pitch on FSDP is kinda of setting the expectation that with the expanding the number of gpus the memory footprint kind of scale down linearly, this happens in models such as DeepVit.

a- I wonder if we are going to consider the world-size in our calculations, that how many all_gather can kind of meet the expectation.

b- it would be great if we could benchmark the impact of current all_gathers vs reducing the number of all_gathers and  increasing the batch size on the performance.

c- FSDP kind of brings a nice simplicity for users, it is nice that its light weight on configs and maybe exposing this knob #all_gather, needs a very deep knowledge of users on internals of disturbed stack and then its very experimental for them to tune. I wonder if using FSDP unit size to calculate the number of required all_gather in respect with world_size might be something to consider.

d- Also, on DeepVit it seems memory get reduced as expected, which may tell something about the FSDP unit size/ potential communication overhead that could be optimized by tuning the number of all_gathers based on the size/ communication time.

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -692,6 +694,43 @@ def next_iter(self):
                 self.warn_status = _ExecOrderWarnStatus.WARNED
 
 
+class FreeEventQueue:
```
make it private _FreeEventQueue?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -692,6 +694,43 @@ def next_iter(self):
+        """Dequeues a free event if possible."""
+        if self._queue:
+            event = self._queue.popleft()
+            return event
```
nit: just `return self._queue.popleft()`?

awgu replied:
It appears that `popleft()` cannot take a default argument to return if the queue is empty, and instead, it will raise `IndexError` if the queue is empty. In that case, I believe we need to check non-emptiness before calling `popleft()`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -894,7 +932,12 @@ class FullyShardedDataParallel(nn.Module):
+        limit_all_gathers (bool): If ``False``, then FSDP allows the CPU
+            thread to schedule all-gathers without any extra synchronization.
+            If ``True``, then FSDP explicitly synchronizes the CPU thread to
+            prevent too many in-flight all-gathers. This ``bool`` only affects
```
should we note that this is 2 at the moment? Do we want to expose any knob for further tuning by advanced users?

awgu replied:
At the moment, I do not want to expose the knob (see the updated PR summary). Given that we do not expose the knob, I am not sure if it is valuable to explain the internal limiter implementation to the user.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -909,6 +952,7 @@ def __init__(
         param_init_fn: Optional[Callable[[nn.Module], None]] = None,
         device_id: Optional[Union[int, torch.device]] = None,
         sync_module_states: bool = False,
+        limit_all_gathers: bool = True,
```
default seems inconsistent with the doc?

awgu replied:
Good catch! This is a mistake.

rohan-varma reviewed (commented):
LGTM overall, a few questions / clarifications. Looking forward to the detailed experiment results!

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.





[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

awgu commented:
@pytorchbot rebase -s

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

Commit:
Update on "[FSDP] Add rate limiter"


**Overview**
This PR adds a `bool` argument `limit_all_gathers` to the FSDP constructor, defaulted to `False`.
- Setting `limit_all_gathers=True` limits the max number of inflight all-gathers to 2 (an empirically chosen constant), preventing a fast CPU thread from over-allocating blocks to the all-gather stream.
- When experiencing a high number of CUDA malloc retries, the limiter can help reduce the number and hence lead to QPS improvement.

**Exploration**
I experimented with both a count-based limiter and size-based limiter (where the size is based on the inflight all-gather size in bytes).
- The size-based limiter did not provide any advantage, only confusing the developer and user alike on what threshold to set.
- For the count-based approach, I decided not to expose the max number of inflight all-gathers to the user since values other than 2 do not show improvements and exposing the knob may confuse users.

**T5-11B**
T5-11B evidences the performance gain from enabling the limiter and that a limit of 2 is a reasonable choice. This is run on an AWS cluster with 8 A100s per node and EFA. For both 2 and 4 nodes, we scale the batch size maximally before hitting OOM, which is a common practice. 

<p float="left">
  <img src="https://user-images.githubusercontent.com/31054793/188936036-04427da9-f492-4e50-9b35-ff64665d9815.png" width="400" />
  <img src="https://user-images.githubusercontent.com/31054793/188936045-f44e659f-1e18-4ea7-8c78-0fce4ff8fb48.png" width="400" /> 
</p>

For 2 nodes, the limit of 2 yields 3.01x QPS improvement, and for 4 nodes, the limit of 2 yields 2.87x QPS improvement.

We need more data points, but the limiter may simplify the batch size scaling workflow. Normally, a practitioner may scale until hitting OOM and back off until there are few CUDA malloc retries. However, now the practitioner may be able to scale until hitting OOM and simply turn on the limiter to reduce the number of retries instead of backing off.

Differential Revision: [D39331201](https://our.internmc.facebook.com/intern/diff/D39331201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331201).

awgu commented:
@pytorchbot merge

awgu commented:
@pytorchbot merge
